File: src/main/java/com/github/solairerove/algs4/leprosorium/data_structures/binary_search_tree/BST.java
package com.github.solairerove.algs4.leprosorium.data_structures.binary_search_tree;

import edu.princeton.cs.algs4.Queue;
import edu.princeton.cs.algs4.StdOut;

public class BST<Key extends Comparable<Key>, Value> {
    private Node root;

    private class Node {
        private Key key;
        private Value value;
        private Node left, right;
        private int n; // количество узлов в поддереве с этим корнем

        public Node(Key key, Value value, int n) {
            this.key = key;
            this.value = value;
            this.n = n;
        }
    }

    public int size() {
        return size(root);
    }

    private int size(Node x) {
        if (x == null) return 0;
        else return x.n;
    }

    public Value get(Key key) {
        return get(root, key);
    }

    private Value get(Node x, Key key) {
        if (x == null) return null;

        int cmp = key.compareTo(x.key);
        if (cmp < 0) return get(x.left, key);
        else if (cmp > 0) return get(x.right, key);
        else return x.value;
    }

    public void put(Key key, Value value) {
        root = put(root, key, value);
    }

    private Node put(Node x, Key key, Value value) {
        if (x == null) return new Node(key, value, 1);

        int cmp = key.compareTo(x.key);
        if (cmp < 0) x.left = put(x.left, key, value);
        else if (cmp > 0) x.right = put(x.right, key, value);
        else x.value = value;

        x.n = size(x.left) + size(x.right) + 1;
        return x;
    }

    public Key min() {
        return min(root).key;
    }

    private Node min(Node x) {
        if (x.left == null) return x;
        return min(x.left);
    }

    public Key max() {
        return max(root).key;
    }

    private Node max(Node x) {
        if (x.right == null) return x;
        return max(x.right);
    }

    public Key floor(Key key) {
        Node x = floor(root, key);
        if (x == null) return null;
        else return x.key;
    }

    private Node floor(Node x, Key key) {
        if (x == null) return null;

        int cmp = key.compareTo(x.key);
        if (cmp == 0) return x;
        if (cmp < 0) return floor(x.left, key);

        Node t = floor(x.right, key);
        if (t != null) return t;
        else return x;
    }

    public Key select(int k) {
        return select(root, k).key;
    }

    private Node select(Node x, int k) {
        if (x == null) return null;

        int t = size(x.left);
        if (t > k) return select(x.left, k);
        else if (t < k) return select(x.right, k - t - 1);
        else return x;
    }

    public int rank(Key key) {
        return rank(root, key);
    }

    private int rank(Node x, Key key) {
        if (x == null) return 0;

        int cmp = key.compareTo(x.key);
        if (cmp < 0) return rank(x.left, key);
        else if (cmp > 0) return 1 + size(x.left) + rank(x.right, key);
        else return size(x.left);
    }

    public void deleteMin() {
        root = deleteMin(root);
    }

    private Node deleteMin(Node x) {
        if (x.left == null) return x.right;

        x.left = deleteMin(x.left);

        x.n = size(x.left) + size(x.right) + 1;
        return x;
    }

    public void delete(Key key) {
        root = delete(root, key);
    }

    private Node delete(Node x, Key key) {
        if (x == null) return null;

        int cmp = key.compareTo(x.key);
        if (cmp < 0) x.left = delete(x.left, key);
        else if (cmp > 0) x.right = delete(x.right, key);
        else {
            if (x.right == null) return x.left;
            if (x.left == null) return x.right;
            Node t = x;
            x = min(t.right);
            x.right = deleteMin(t.right);
            x.left = t.left;
        }

        x.n = size(x.left) + size(x.right) + 1;
        return x;
    }

    public void print() {
        print(root);
    }

    private void print(Node x) {
        if (x == null) return;

        print(x.left);
        StdOut.printf("%s ", x.key);
        print(x.right);
    }

    public Iterable<Key> keys() {
        return keys(min(), max());
    }

    public Iterable<Key> keys(Key low, Key high) {
        Queue<Key> queue = new Queue<>();
        keys(root, queue, low, high);

        return queue;
    }

    private void keys(Node x, Queue<Key> queue, Key low, Key high) {
        if (x == null) return;

        int cmpLow = low.compareTo(x.key);
        int cmpHigh = high.compareTo(x.key);
        if (cmpLow < 0) keys(x.left, queue, low, high);
        if (cmpLow <= 0 && cmpHigh >= 0) queue.enqueue(x.key);
        if (cmpHigh > 0) keys(x.right, queue, low, high);
    }
}

File: src/main/kotlin/com/github/solairerove/algs4/leprosorium/symbol_table/BST.kt
package com.github.solairerove.algs4.leprosorium.symbol_table

import java.util.*

fun main() {
    val bst = BST<String, Int>()
    bst.put("S", 0)
    bst.put("E", 1)
    bst.put("A", 2)
    bst.put("R", 3)
    bst.put("C", 4)
    bst.put("H", 5)
    bst.put("E", 6)
    bst.put("X", 7)
    bst.put("A", 8)
    bst.put("M", 9)
    bst.put("P", 10)
    bst.put("L", 11)
    bst.put("E", 12)

    println(bst.size()) // 10
    println(bst.get("E")) // 12
    println(bst.min()) // A
    println(bst.max()) // X
    println(bst.floor("G")) // E
    println(bst.ceiling("N")) // M
    println(bst.select(3)) // H
    println(bst.rank("H")) // 3
    bst.deleteMin()
    println(bst.min()) // C
    bst.delete("E")
    println(bst.get("E")) // null
    println("-----")
    bst.print()
    println("-----")
    println(bst.keys())
}

class BST<Key : Comparable<Key>, Value> {
    private var root: Node? = null

    inner class Node(var key: Key, value: Value, var n: Int) {
        var value: Value? = value
        var right: Node? = null
        var left: Node? = null
    }

    fun size(): Int = size(x = root)

    private fun size(x: Node?): Int {
        if (x == null) return 0
        return x.n
    }

    fun get(key: Key): Value? = get(x = root, key = key)

    private fun get(x: Node?, key: Key): Value? {
        if (x == null) return null

        val cmp = key.compareTo(x.key)
        return when {
            cmp < 0 -> get(x.left, key)
            cmp > 0 -> get(x.right, key)
            else -> x.value
        }
    }

    fun put(key: Key, value: Value) {
        root = put(x = root, key = key, value = value)
    }

    private fun put(x: Node?, key: Key, value: Value): Node {
        if (x == null) return Node(key = key, value = value, n = 1)

        val cmp = key.compareTo(x.key)
        when {
            cmp < 0 -> x.left = put(x = x.left, key = key, value = value)
            cmp > 0 -> x.right = put(x = x.right, key = key, value = value)
            else -> x.value = value
        }
        x.n = size(x.left) + size(x.right) + 1

        return x
    }

    fun min(): Key = min(root!!).key

    private fun min(x: Node): Node {
        if (x.left == null) return x
        return min(x.left!!)
    }

    fun max(): Key = max(root!!).key

    private fun max(x: Node): Node {
        if (x.right == null) return x
        return max(x.right!!)
    }

    fun floor(key: Key): Key? {
        val x = floor(root, key) ?: return null
        return x.key
    }

    private fun floor(x: Node?, key: Key): Node? {
        if (x == null) return null

        val cmp = key.compareTo(x.key)
        if (cmp == 0) return x
        if (cmp < 0) return floor(x.left, key)

        return floor(x.right, key) ?: x
    }

    fun ceiling(key: Key): Key? {
        val x = ceiling(root, key) ?: return null
        return x.key
    }

    private fun ceiling(x: Node?, key: Key): Node? {
        if (x == null) return null

        val cmp = key.compareTo(x.key)
        if (cmp == 0) return x
        if (cmp < 0) return ceiling(x.left, key) ?: x

        return floor(x.right, key)
    }

    fun rank(key: Key): Int = rank(key, root)

    private fun rank(key: Key, x: Node?): Int {
        if (x == null) return 0

        val cmp = key.compareTo(x.key)
        return when {
            cmp < 0 -> rank(key, x.left)
            cmp > 0 -> rank(key, x.right) + size(x.left) + 1
            else -> return size(x.left)
        }
    }

    fun select(k: Int): Key? = select(root, k)?.key

    private fun select(x: Node?, k: Int): Node? {
        if (x == null) return null

        val t = size(x.left)
        return when {
            t > k -> select(x.left, k)
            t < k -> select(x.right, k - t - 1)
            else -> x
        }
    }

    fun deleteMin() {
        root = deleteMin(root)
    }

    private fun deleteMin(x: Node?): Node? {
        if (x?.left == null) return x?.right
        x.left = deleteMin(x.left)
        x.n = size(x.left) + size(x.right) + 1
        return x
    }

    fun delete(key: Key) {
        root = delete(root, key)
    }

    private fun delete(node: Node?, key: Key): Node? {
        var x = node ?: return null

        val cmp = key.compareTo(x.key)
        when {
            cmp < 0 -> x.left = delete(x.left, key)
            cmp > 0 -> x.right = delete(x.right, key)
            else -> {
                if (x.right == null) return x.left
                if (x.left == null) return x.right

                val t: Node = x
                x = min(t.right!!)
                x.right = deleteMin(t.right)
                x.left = t.left
            }
        }
        x.n = size(x.left) + size(x.right) + 1
        return x
    }

    fun print() {
        print(root)
    }

    private fun print(x: Node?) {
        if (x == null) return
        print(x.left)
        println(x.key)
        print(x.right)
    }

    fun keys(): Iterable<Key> = keys(min(), max())

    fun keys(low: Key, high: Key): Iterable<Key> {
        val queue = LinkedList<Key>()
        keys(root, queue, low, high)
        return queue
    }

    private fun keys(x: Node?, queue: Queue<Key>, low: Key, high: Key) {
        if (x == null) return
        val cmpLow = low.compareTo(x.key)
        val cmpHigh = high.compareTo(x.key)
        if (cmpLow < 0) keys(x.left, queue, low, high)
        if (cmpLow <= 0 && cmpHigh >= 0) queue.add(x.key)
        if (cmpHigh > 0) keys(x.right, queue, low, high)
    }
}
